### DISCUSSION TITLE
Collection Of Specific Objects With Empty Collection Not Allowed?

### DISCUSSION THREAD

**Discussion by TheFerrettDeveloper** [spatie/laravel-data#765] [Q&A]
So I have this field on a data object: 

`        /** @var Collection <int, SearchTerms> */`
`        public Collection $search_terms`

Yet whenever I initiate it with an empty Collection (as not every search has these terms), I'm getting an error that "The search terms field is required".  

Is that... expected behavior?  Is there a way to say "It's a Collection of these objects, but no objects is also okay?"

**Comment by rubenvanassche** [comment]
Use a default?

```php
public Collection $search_terms = new Collection()
```

Only allowed in constructors if I'm not mistaken

**Reply by cyppe** [reply]
Doing this still does not pass validation if you pass empty array 

somethingData::from(['search_terms' = []]);

Or even skip it.

It *seems* to work with adding 

```
  #[Sometimes]
  public Collection $search_terms = new Collection()
```

But will test a bit more.  But @rubenvanassche do you have some input regarding what is most "correct" way of doing this?

I do not want to use Collection|Optional as type, as phpstan on high level will not accept doing a check like:

$object->search_terms->isNotEmpty() as it might be of type Optional.

**Reply by cyppe** [reply]
Did some more testing, and it seems like everything works as expected out of the box as long as I use a collection of Data objects:

```
        /** @var Collection<int, ProductVariantData> */
        public Collection $variants,
````

I only face issue when I try to to store some kind of array inside the Collection.   So I just switched the remaining Collection properties to use Data objects and it works fine now where it auto creates empty collection when I send empty array.
